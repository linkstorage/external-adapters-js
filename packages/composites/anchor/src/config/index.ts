import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config as DefaultConfig } from '@chainlink/types'

export const DEFAULT_ENDPOINT = 'price'
export const FIXED_POINT_DECIMALS = 18
export const NAME = 'ANCHOR'

export const DEFAULT_ANCHOR_VAULT_CONTRACT_ADDRESS = '0xA2F987A546D4CD1c607Ee8141276876C26b72Bdf'
export const DEFAULT_TERRA_BLUNA_CONTRACT_ADDRESS = 'terra1mtwph2juhj0rvjz7dy92gvl6xvukaxu8rfv8ts'
export const DEFAULT_STETH_POOL_CONTRACT_ADDRESS = '0xdc24316b9ae028f1497c275eb9192a3ea0f67022'

// Bombay
// TODO:  This is only for testing right now.  Make sure to update these to the Columbus addresses once they have been deployed
export const DEFAULT_LUNA_TERRA_FEED_ADDRESS = 'terra1u475ps69rmhpf4f4gx2pc74l7tlyu4hkj4wp9d'
export const DEFAULT_ETH_TERRA_FEED_ADDRESS = 'terra1dw5ex5g802vgrek3nzppwt29tfzlpa38ep97qy'
export const DEFAULT_BTC_TERRA_FEED_ADDRESS = 'terra134m32c6p87df4cjx36vhgxnlhf06pfgvj7jsx7'

export interface Config extends DefaultConfig {
  anchorVaultContractAddress: string
  terraBLunaContractAddress: string
  stEthPoolContractAddress: string
  feedAddresses: {
    [T: string]: string
  }
}

export const makeConfig = (prefix?: string): Config => {
  return {
    ...Requester.getDefaultConfig(prefix),
    rpcUrl: util.getRequiredEnv('RPC_URL', prefix),
    defaultEndpoint: util.getEnv('API_ENDPOINT', prefix) || DEFAULT_ENDPOINT,
    anchorVaultContractAddress:
      util.getEnv('ANCHOR_VAULT_CONTRACT_ADDRESS', prefix) || DEFAULT_ANCHOR_VAULT_CONTRACT_ADDRESS,
    terraBLunaContractAddress:
      util.getEnv('TERRA_BLUNA_CONTRACT_ADDRESS', prefix) || DEFAULT_TERRA_BLUNA_CONTRACT_ADDRESS,
    stEthPoolContractAddress:
      util.getEnv('STETH_POOL_CONTRACT_ADDRESS', prefix) || DEFAULT_STETH_POOL_CONTRACT_ADDRESS,
    feedAddresses: {
      luna: util.getEnv('LUNA_TERRA_FEED_ADDRESS', prefix) || DEFAULT_LUNA_TERRA_FEED_ADDRESS,
      eth: util.getEnv('ETH_TERRA_FEED_ADDRESS', prefix) || DEFAULT_ETH_TERRA_FEED_ADDRESS,
      btc: util.getEnv('BTC_TERRA_FEED_ADDRESS', prefix) || DEFAULT_BTC_TERRA_FEED_ADDRESS,
    },
  }
}
