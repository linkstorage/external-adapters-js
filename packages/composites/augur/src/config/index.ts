import { Config as DefaultConfig } from '@chainlink/types'
import { ethers } from 'ethers'
import { HTTP, util } from '@chainlink/ea-bootstrap'

export type Config = DefaultConfig & {
  wallet: ethers.Wallet
}

export const NAME = 'AUGUR'

export const makeConfig = (prefix?: string): Config => {
  const rpcUrl = util.getRequiredEnvWithFallback('ETHEREUM_RPC_URL', ['RPC_URL'], prefix)
  const privateKey = util.getRequiredEnv('PRIVATE_KEY', prefix)
  const provider = new ethers.providers.JsonRpcProvider(rpcUrl)
  const wallet = new ethers.Wallet(privateKey, provider)

  return {
    ...HTTP.getDefaultConfig(prefix),
    verbose: true,
    wallet,
  }
}
