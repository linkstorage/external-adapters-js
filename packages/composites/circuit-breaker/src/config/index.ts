import { HTTP } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/types'

export const NAME = 'CIRCUIT_BREAKER'

export const makeConfig = (prefix = ''): Config => {
  return HTTP.getDefaultConfig(prefix)
}
