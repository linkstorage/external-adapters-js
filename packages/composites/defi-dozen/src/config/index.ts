import { HTTP } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/types'

export const NAME = 'DEFI_DOZEN'

export const makeConfig = (prefix?: string): Config => {
  return HTTP.getDefaultConfig(prefix)
}
