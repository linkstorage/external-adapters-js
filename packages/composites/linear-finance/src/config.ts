import { Config as BaseConfig } from '@chainlink/types'
import { HTTP } from '@chainlink/ea-bootstrap'
import xbci from './indices/xbci'
import xlci from './indices/xlci'

export const INDICES = ['xbci', 'xlci']
export type IndexType = typeof INDICES[number]

export type Config = BaseConfig & {
  indices: {
    [key in IndexType]: string
  }
}

export const makeConfig = (prefix?: string): Config => {
  return {
    ...HTTP.getDefaultConfig(prefix),
    indices: {
      xbci,
      xlci,
    },
  }
}
