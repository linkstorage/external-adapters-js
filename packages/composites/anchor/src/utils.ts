import { AdapterContext, AdapterRequest, AdapterResponse } from '@chainlink/types'
import * as TA from '@chainlink/token-allocation-adapter'
import { BigNumber, ethers } from 'ethers'
import { FIXED_POINT_DECIMALS } from './config'

export const getTokenPrice = async (
  input: AdapterRequest,
  context: AdapterContext,
  symbol: string,
  decimals = 18,
): Promise<AdapterResponse> => {
  const _config = TA.makeConfig()
  const _execute = TA.makeExecute(_config)
  const allocations = [
    {
      symbol,
      balance: BigNumber.from(10).pow(decimals).toString(),
      decimals,
    },
  ]
  return await _execute({ id: input.id, data: { ...input.data, allocations } }, context)
}

export const convertUSDQuote = async (
  input: AdapterRequest,
  context: AdapterContext,
  usdPrice: ethers.BigNumber,
  targetQuote: string,
  targetQuoteDecimals = 18,
): Promise<ethers.BigNumber> => {
  const targetQuoteUSDRateResp = await getTokenPrice(
    input,
    context,
    targetQuote,
    targetQuoteDecimals,
  )
  const targetQuoteUSDRate = ethers.utils.parseUnits(
    targetQuoteUSDRateResp.data.result.toString(),
    FIXED_POINT_DECIMALS,
  )
  return usdPrice.mul(BigNumber.from(10).pow(FIXED_POINT_DECIMALS)).div(targetQuoteUSDRate)
}
