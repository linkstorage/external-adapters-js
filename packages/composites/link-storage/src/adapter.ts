import { Requester, Validator, Logger } from '@chainlink/ea-bootstrap'
import {
  AdapterContext,
  AdapterRequest,
  AdapterResponse,
  Config,
  Execute,
  ExecuteWithConfig,
  ExecuteFactory,
  InputParameters,
} from '@chainlink/types'
import { create, IPFSHTTPClient } from 'ipfs-http-client'
import { base16 } from 'multiformats/bases/base16'
import { makeConfig } from './config'

export const inputParameters: InputParameters = {
  data: {
    required: true,
    description: 'The data to write',
  },
  destChain: {
    aliases: ['targetChain'],
    description: 'The target chain id',
  },
  contractAddr: {
    aliases: ['targetAddress', 'contractAddress'],
    description: 'The contract address will be interacted',
  },
  func: {
    aliases: ['function'],
    description: 'Function to be called',
    required: false,
  },
  cidVersion: {
    required: false,
    description: 'The CID version to be returned',
    type: 'number',
    default: 1,
  },
}

export const execute: ExecuteWithConfig<Config> = async (
  request: AdapterRequest,
  _: any,
  config: any,
): Promise<AdapterResponse> => {
  const validator = new Validator(request, inputParameters)
  Logger.info('processing:', validator.validated.data)
  Logger.info('config data:', config)

  if (validator.error) throw validator.error

  const jobRunID = validator.validated.jobRunID
  const data = validator.validated.data.data
  const cidVersion = validator.validated.data.cidVersion

  const projectId = config.projectId
  const projectSecret = config.projectSecret
  if (config.ipfsURL.includes('infura') && (!projectId || !projectSecret)) {
    throw Error('You MUST config the infura project id and secret')
  }

  const auth =
    projectId && projectSecret
      ? `Basic ${Buffer.from(projectId + ':' + projectSecret).toString('base64')}`
      : ''

  const client = create({
    url: config.ipfsURL,
    headers: {
      authorization: auth,
    },
  })
  const options = { cidVersion, hashAlg: 'sha3-224' }

  const cid = await putFile(serialize({ jobRunID, ...data, ...options }), client, options)
  Logger.debug('cid:', cid.toString())

  const base16HexCid = `0x${cid.toString(base16).slice(1)}`
  const response = { data: { result: base16HexCid, cid: cid.toString() } }

  return Requester.success(jobRunID, response, config.verbose)
}

export const makeExecute: ExecuteFactory<Config> = (config?: Config): Execute => {
  return async (request: AdapterRequest, context: AdapterContext) =>
    execute(request, context, config || makeConfig())
}

const putFile = async (
  data: string | Uint8Array,
  client: IPFSHTTPClient,
  options: Record<string, unknown>,
) => {
  const { cid } = await client.add(data, options)
  return cid
}

const serialize = (data: string | object): string | Uint8Array => {
  if (typeof data === 'string') return data

  return Buffer.from(JSON.stringify(data))
}
