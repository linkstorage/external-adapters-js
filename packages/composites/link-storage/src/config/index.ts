import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/types'
import 'dotenv/config'
import { CustomConfig } from '../index.d'

export const NAME = 'LINKSTORAGE'
export const DEFAULT_IPFS_URL = 'http://localhost:5001'

export const makeConfig = (prefix?: string): Config => {
  const config: CustomConfig = Requester.getDefaultConfig(prefix)
  config.ipfsURL = util.getEnv('IPFS_URL') || DEFAULT_IPFS_URL
  config.projectId = util.getEnv('INFURA_PROJECT_ID') || ''
  config.projectSecret = util.getEnv('INFURA_PROJECT_SECRET') || ''
  config.verbose = util.parseBool(util.getEnv('VERBOSE')) || true
  return config
}
