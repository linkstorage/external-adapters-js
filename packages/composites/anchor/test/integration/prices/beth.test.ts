import { AdapterRequest } from '@chainlink/types'
import { server as startServer } from '../../../src'
import nock from 'nock'
import http from 'http'
import request, { SuperTest, Test } from 'supertest'
import { mockBTCUSDPrice, mockETHUSDPrice, mockSTEthUSDPrice } from '../fixtures'

import { ethers, BigNumber } from 'ethers'
import '@chainlink/terra-view-function-adapter'
import { AddressInfo } from 'net'

const mockBigNum = BigNumber.from(10).pow(18)
const mockStETHETHPrice = BigNumber.from('1035144096528344468')

jest.mock('@chainlink/terra-view-function-adapter', () => {
  return {
    ...jest.requireActual('@chainlink/terra-view-function-adapter'),
    makeExecute: jest.fn().mockReturnValue(
      jest.fn().mockReturnValue({
        jobRunID: '1',
        result: {
          round_id: 314711,
          answer: '262009859746',
          started_at: 1645564682,
          updated_at: 1645564682,
          answered_in_round: 314711,
        },
        statusCode: 200,
        data: {
          result: {
            round_id: 314711,
            answer: '262009859746',
            started_at: 1645564682,
            updated_at: 1645564682,
            answered_in_round: 314711,
          },
        },
      }),
    ),
  }
})

jest.mock('ethers', () => {
  const actualModule = jest.requireActual('ethers')
  return {
    ...actualModule,
    ethers: {
      ...actualModule.ethers,
      providers: {
        JsonRpcProvider: function (_: string): ethers.provider.JsonRpcProvider {
          return {}
        },
      },
      Contract: function () {
        return {
          get_rate: (____: string) => {
            return mockBigNum
          },
          get_virtual_price: () => {
            return mockStETHETHPrice
          },
        }
      },
    },
  }
})

let oldEnv: NodeJS.ProcessEnv

describe('price-beth', () => {
  let server: http.Server
  let req: SuperTest<Test>

  beforeAll(async () => {
    oldEnv = JSON.parse(JSON.stringify(process.env))
    server = await startServer()
    req = request(`localhost:${(server.address() as AddressInfo).port}`)
    process.env.COLUMBUS_5_RPC_URL = 'fake-columbus-rpc'
    process.env.API_KEY = 'test'
    process.env.ANCHOR_VAULT_CONTRACT_ADDRESS = 'test-address'
    process.env.COINGECKO_ADAPTER_URL = 'http://localhost:5000'
    process.env.RPC_URL = 'test-rpc-url'

    if (process.env.RECORD) {
      nock.recorder.rec()
    }
  })

  afterAll((done) => {
    if (process.env.RECORD) {
      nock.recorder.play()
    }
    process.env = oldEnv
    nock.restore()
    nock.cleanAll()
    nock.enableNetConnect()
    server.close(done)
  })

  describe('successful calls', () => {
    const jobID = '1'

    it('return success when fetching the USD/BEth price', async () => {
      mockSTEthUSDPrice()
      const data: AdapterRequest = {
        id: jobID,
        data: {
          from: 'BETH',
          to: 'USD',
          source: 'coingecko',
        },
      }

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })

    it('returns success when fetching the ETH/BEth price', async () => {
      mockSTEthUSDPrice()
      mockETHUSDPrice()
      const data: AdapterRequest = {
        id: jobID,
        data: {
          from: 'BETH',
          to: 'ETH',
          source: 'coingecko',
        },
      }

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })

    it('returns success when fetching the BTC/BEth price', async () => {
      mockSTEthUSDPrice()
      mockBTCUSDPrice()

      const data: AdapterRequest = {
        id: jobID,
        data: {
          from: 'BETH',
          to: 'BTC',
          source: 'coingecko',
          quoteDecimals: 8,
        },
      }

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })
})
