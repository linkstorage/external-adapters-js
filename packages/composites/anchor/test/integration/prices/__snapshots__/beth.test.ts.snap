// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`price-beth successful calls return success when fetching the USD/BEth price 1`] = `
Object {
  "data": Object {
    "result": "25311438342229448966",
  },
  "jobRunID": "1",
  "result": "25311438342229448966",
  "statusCode": 200,
}
`;

exports[`price-beth successful calls returns success when fetching the BTC/BEth price 1`] = `
Object {
  "data": Object {
    "result": "966049077953291358",
  },
  "jobRunID": "1",
  "result": "966049077953291358",
  "statusCode": 200,
}
`;

exports[`price-beth successful calls returns success when fetching the ETH/BEth price 1`] = `
Object {
  "data": Object {
    "result": "966049077953291358",
  },
  "jobRunID": "1",
  "result": "966049077953291358",
  "statusCode": 200,
}
`;
